src/loaders/workspace_downloader/workspace_downloader.py

```diff
@@ -0,0 +1,356 @@
+"""
+import argparse
+import itertools
+import json
+import os
+import shutil
+import stat
+import time
+from multiprocessing import Pool, Queue, cpu_count
+
+import docker
+
+from src.clients.AssemblyUtilClient import AssemblyUtil
+from src.clients.workspaceClient import Workspace
+from src.loaders.common import loader_common_names, loader_helper
+
+# setup KB_AUTH_TOKEN as env or provide a token_filename in --token_filename
+# export KB_AUTH_TOKEN="your-kb-auth-token"
+
+# supported source of data
+SOURCE = "WS"
+# filtering applied to list objects
+FILTER_OBJECTS_NAME_BY = "KBaseGenomeAnnotations.Assembly"
+# callback_port
+CALLBACK_PORT = 9999
+
+
+class Conf:
+    def __init__(self, job_dir, output_dir, workers, kb_base_url, token_filename):
+        self.setup_callback_server_envs(job_dir, kb_base_url, token_filename)
+        time.sleep(1)
+        self.start_callback_server(docker.from_env(), "cb")
+        time.sleep(2)
+
+        # os.environ['SDK_CALLBACK_URL'] = self.cb.callback_url
+        token = os.environ["KB_AUTH_TOKEN"]
+        ws_url = os.path.join(kb_base_url, "ws")
+        callback_url = "http://" + loader_helper.get_ip() + ":" + str(CALLBACK_PORT)
+        print("callback_url:", callback_url)
+
+        self.ws = Workspace(ws_url, token=token)
+        self.asu = AssemblyUtil(callback_url, token=token)
+        self.queue = Queue()
+        self.pth = output_dir
+        self.job_dir = job_dir
+        self.pools = Pool(workers, process_input, [self])
+
+    def setup_callback_server_envs(self, job_dir, kb_base_url, token_filename):
+        # initiate env and vol
+        self.env = {}
+        self.vol = {}
+
+        # setup envs required for docker container
+        token = os.environ.get("KB_AUTH_TOKEN")
+        if not token:
+            if not token_filename:
+                raise ValueError("Need to provide a token_filename")
+            token = loader_helper.get_token(token_filename)
+        os.environ["KB_AUTH_TOKEN"] = token
+
+        # used by the callback server
+        self.env["KB_AUTH_TOKEN"] = token
+        # used by the callback server
+        self.env["KB_BASE_URL"] = kb_base_url
+        # used by the callback server
+        self.env["JOB_DIR"] = job_dir
+        # used by the callback server
+        self.env["CALLBACK_PORT"] = CALLBACK_PORT
+
+        # setup volumes required for docker container
+        docker_host = os.environ["DOCKER_HOST"]
+        if docker_host.startswith("unix:"):
+            docker_host = docker_host[5:]
+        self.vol[job_dir] = {"bind": job_dir, "mode": "rw"}
+        self.vol[docker_host] = {"bind": "/run/docker.sock", "mode": "rw"}
+
+    def start_callback_server(self, client, container_name):
+        container = None
+        try:
+            container = client.containers.get(container_name)
+        except Exception as e:
+            print("container does not exist and will fetch scanon/callback ")
```

Comment:
We need to get this into a kbase namespace, preferably built by GHA. Can you make it a constant in the common names file for now, so it's easy to change later?

Reply:
now start a new callback server container every time with diff name

Reply:
I'm talking about the image name, `scanon/callback`. That should be a constant so it's easy to change when we move the image to GHA

Reply:
Can you make the name CALLBACK_IMAGE_NAME or something like that?

Reply:
Updated!

Reply:
:+1: